applicationset/services/pull_request/azure_devops.go

```diff
@@ -0,0 +1,137 @@
+package pull_request
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/microsoft/azure-devops-go-api/azuredevops"
+	core "github.com/microsoft/azure-devops-go-api/azuredevops/core"
+	git "github.com/microsoft/azure-devops-go-api/azuredevops/git"
+)
+
+const AZURE_DEVOPS_DEFAULT_URL = "https://dev.azure.com"
+
+type AzureDevOpsClientFactory interface {
+	// Returns an Azure Devops Client interface.
+	GetClient(ctx context.Context) (git.Client, error)
+}
+
+type devopsFactoryImpl struct {
+	connection *azuredevops.Connection
+}
+
+func (factory *devopsFactoryImpl) GetClient(ctx context.Context) (git.Client, error) {
+	gitClient, err := git.NewClient(ctx, factory.connection)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get new Azure DevOps git client for SCM generator: %w", err)
+	}
+	return gitClient, nil
+}
+
+type AzureDevOpsService struct {
+	clientFactory AzureDevOpsClientFactory
+	project       string
+	repo          string
+	labels        []string
+}
+
+var _ PullRequestService = (*AzureDevOpsService)(nil)
+var _ AzureDevOpsClientFactory = &devopsFactoryImpl{}
+
+func NewAzureDevOpsService(ctx context.Context, token, url, organization, project, repo string, labels []string) (PullRequestService, error) {
+	organizationUrl := buildURL(url, organization)
+
+	var connection *azuredevops.Connection
+	if token == "" {
+		connection = azuredevops.NewAnonymousConnection(organizationUrl)
+	} else {
+		connection = azuredevops.NewPatConnection(organizationUrl, token)
+	}
+
+	return &AzureDevOpsService{
+		clientFactory: &devopsFactoryImpl{connection: connection},
+		project:       project,
+		repo:          repo,
+		labels:        labels,
+	}, nil
+}
+
+func (a *AzureDevOpsService) List(ctx context.Context) ([]*PullRequest, error) {
+
+	client, err := a.clientFactory.GetClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Azure DevOps client: %w", err)
+	}
+
+	args := git.GetPullRequestsByProjectArgs{
+		Project:        &a.project,
+		SearchCriteria: &git.GitPullRequestSearchCriteria{},
+	}
+
+	azurePullRequests, err := client.GetPullRequestsByProject(ctx, args)
+
+	if err != nil {
+		return nil, err
+	}
+
+	pullRequests := []*PullRequest{}
+
+	for _, pr := range *azurePullRequests {
+		azureDevOpsLabels := convertLabels(pr.Labels)
+		if !containAzureDevOpsLabels(a.labels, azureDevOpsLabels) {
+			continue
+		}
+		if *pr.Repository.Name == a.repo {
+			pullRequests = append(pullRequests, &PullRequest{
+				Number:  *pr.PullRequestId,
+				Branch:  strings.Replace(*pr.SourceRefName, "refs/heads/", "", 1),
+				HeadSHA: *pr.LastMergeSourceCommit.CommitId,
+				Labels:  azureDevOpsLabels,
+			})
+		}
+	}
```

Comment:
There are a few unprotected pointer de-references here. Would you mind checking for nil and either throwing an error or coercing to a zero value (i.e. `""`)?

Reply:
Added nil checks for pointers.